### PR TITLE
fix #1261, #1127, replace spray with akka-http

### LINKED ISSUE #1261: Replace Spray with Akka-HTTP
Reason:
1. First, development of spray is merged to akka-http.
2. akka-http has flow control. to avoid OOM
3. akka-http support streaming large files.

### PR DESCRIPTION
**Pull request by clockfly**

### CODE REVIEW

**Comment by clockfly** [comment]
1. spray web socket is removed, as akka-http will have its own websocket support.

File: daemon/src/main/scala/io/gearpump/util/FileServer.scala [hunk 1]
```diff
@@ -15,123 +15,137 @@
...
+    connection.flatMap(_.unbind())
   }
 }
 
 object FileServer {
-  object GetPort
+
+  implicit def filePathFormat: JsonFormat[FilePath] = jsonFormat1(FilePath.apply)
+
+  def newClient = null
```
Comment (huafengw) [inline daemon/src/main/scala/io/gearpump/util/FileServer.scala]: should be deleted?

**Comment by clockfly** [comment]
TimeOutSchedulerSpec UT failure is not related.

File: core/src/main/scala/io/gearpump/jarstore/JarStoreService.scala [hunk 2]
```diff
@@ -35,14 +35,14 @@ trait JarStoreService {
    */
   val scheme : String
 
-  def init(config: Config, actorRefFactory: ActorRefFactory)
+  def init(config: Config, system: ActorSystem)
 
   /**
     * This function will copy the local file to the remote JarStore, called from client side.
    * @param localFile The local file
    * @param remotePath The path on JarStore
```
Comment (manuzhang) [inline core/src/main/scala/io/gearpump/jarstore/JarStoreService.scala]: please remove the doc as well.

File: examples/distributeservice/src/main/scala/io/gearpump/experiments/distributeservice/FileServer.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import io.gearpump.util.FileUtils
+import scala.util.{Failure, Success, Try}
+
+/**
+ *
+ * Should not use this to server too big files(more than 100MB), otherwise OOM may happen.
+ *
+ * port: set port to 0 if you want to bind to random port
+ */
+class FileServer(rootDir: File, host: String, port : Int) extends Actor with Stash {
```
Comment (manuzhang) [inline examples/distributeservice/src/main/scala/io/gearpump/experiments/distributeservice/FileServer.scala]: where is this class used ?

Comment (clockfly) [inline examples/distributeservice/src/main/scala/io/gearpump/experiments/distributeservice/FileServer.scala]: IT is used by project distributeservice.

File: daemon/src/main/scala/io/gearpump/util/FileDirective.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        }
+      }
+    }
+  }
+
+  def downloadFile(file: File): Route = {
+    val responseEntity = HttpEntity(
+      MediaTypes.`application/octet-stream`,
+      file.length,
+      SynchronousFileSource(file, CHUNK_SIZE))
```
Comment (stanleyxu2005) [inline daemon/src/main/scala/io/gearpump/util/FileDirective.scala]: Constant is great. But I mean to configure the size in user conf. Different user might want use/restrict a particular chunk size for their web service.

Comment (clockfly) [inline daemon/src/main/scala/io/gearpump/util/FileDirective.scala]: I don't think there is need for user to config it.

**Comment by huafengw** [comment]
+1

**Pull request merged**